### PR TITLE
[BEAM-886] Some performance improvements to NewDoFn

### PR DESCRIPTION
**Pull request by sb2nov**
- Add types to some of variables for performance
- Do minimal work in the process function by stashing the placeholders to be replaced.

R: @robertwb PTAL
CC: @aaltay

---

Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by sb2nov** [comment]
R: @aaltay PTAL

File: sdks/python/apache_beam/runners/common.py [hunk 1]
```diff
@@ -177,66 +246,37 @@ def old_dofn_process(self, element):
...
-    self_in_args = int(self.dofn.is_process_bounded())
 
     # Call for the process function for each window if has windowed side inputs
     # or if the process accesses the window parameter. We can just call it once
     # otherwise as none of the arguments are changing
-    if self.has_windowed_side_inputs or core.NewDoFn.WindowParam in defaults:
+    if self.has_windowed_side_inputs:
       windows = element.windows
     else:
       windows = [window.GlobalWindow()]
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: Cache [window.GlobalWindow()]

File: sdks/python/apache_beam/runners/common.py [hunk 2]
```diff
@@ -177,66 +246,37 @@ def old_dofn_process(self, element):
...
-            final_args.append(args.next())
-          except StopIteration:
-            if a not in kwargs:
-              kwargs[a] = d
-      final_args.extend(list(args))
-      self._process_outputs(element, self.dofn.process(*final_args, **kwargs))
+        args, kwargs = self.args, self.kwargs
+
+      for i, p in self.placeholders:
+        if p == core.NewDoFn.ElementParam:
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: Use is rather than == for speed.

File: sdks/python/apache_beam/runners/common.py [hunk 3]
```diff
@@ -177,66 +246,37 @@ def old_dofn_process(self, element):
...
+          args[i] = element.value
+        elif p == core.NewDoFn.ContextParam:
+          args[i] = self.context
+        elif p == core.NewDoFn.WindowParam:
+          args[i] = w
+        elif p == core.NewDoFn.TimestampParam:
+          args[i] = element.timestamp
+      self._process_outputs(element, self.dofn.process(*args, **kwargs))
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: Might be worth a special case for not having kwargs (as that is rare and makes the function call more expensive).

File: sdks/python/apache_beam/runners/common.py [hunk 4]
```diff
@@ -177,66 +246,37 @@ def old_dofn_process(self, element):
       self.context.set_element(element)
       self._process_outputs(element, self.dofn_process(self.context))
 
+  def new_dofn_simple_process(self, element):
+    self._process_outputs(element, self.dofn.process(element.value))
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: Consider caching dofn.process (here and elsewhere).

Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: Done

File: sdks/python/apache_beam/runners/common.py (hunk 1 above)
Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: Done

File: sdks/python/apache_beam/runners/common.py (hunk 2 above)
Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: Will `is`  work here? I check the id of the objects and they seemed different.

File: sdks/python/apache_beam/runners/common.py (hunk 3 above)
Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: Done

**Comment by sb2nov** [comment]
Thanks for the feedback. PTAL again @robertwb

File: sdks/python/apache_beam/runners/common.py (hunk 2 above)
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: That's strange. Maybe it's due to the pickle round trip... Add a TODO to investigate.

Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: Done.

File: sdks/python/apache_beam/runners/common.py [hunk 5]
```diff
@@ -177,66 +246,38 @@ def old_dofn_process(self, element):
       self.context.set_element(element)
       self._process_outputs(element, self.dofn_process(self.context))
 
+  def new_dofn_simple_process(self, element):
+    self._process_outputs(element, self.dofn_process(element.value))
+
+  def _new_dofn_window_process(self, element, args, kwargs, w):
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: spell out "window" for the last parameter (and in the pxd too).

File: sdks/python/apache_beam/runners/common.py [hunk 6]
```diff
@@ -177,66 +246,38 @@ def old_dofn_process(self, element):
...
-          # If no more args are present then the value must be passed via kwarg
-          try:
-            final_args.append(args.next())
-          except StopIteration:
-            if a not in kwargs:
-              kwargs[a] = d
-      final_args.extend(list(args))
-      self._process_outputs(element, self.dofn.process(*final_args, **kwargs))
+      self._new_dofn_window_process(
+          element, self.args, self.kwargs, self.global_window)
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: Can you simply pass None rather than self.global_window here?

Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: Can someone access the window param in this case ? or does it make no sense to do that ?

File: sdks/python/apache_beam/runners/common.py (hunk 5 above)
Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: Done.

File: sdks/python/apache_beam/runners/common.py (hunk 6 above)
Comment (sb2nov) [inline sdks/python/apache_beam/runners/common.py]: Done

Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: This is the case where the window isn't being accessed.

**Comment by robertwb** [comment]
LGTM

**Comment by sb2nov** [comment]
Thanks for the review.

**Pull request closed without merging**